client/interfaces.go

```diff
@@ -164,9 +168,20 @@ type ContainerServer interface {
 	CopyStoragePoolVolume(pool string, source ContainerServer, sourcePool string, volume api.StorageVolume, args *StoragePoolVolumeCopyArgs) (op *RemoteOperation, err error)
 	MoveStoragePoolVolume(pool string, source ContainerServer, sourcePool string, volume api.StorageVolume, args *StoragePoolVolumeMoveArgs) (op *RemoteOperation, err error)
 
+	// Cluster functions ("cluster" API extensions)
+	GetCluster(password string) (cluster *api.Cluster, err error)
+	BootstrapCluster(name string) (op *Operation, err error)
+	AcceptNode(targetPassword, name, address string, schema, api int, pools []api.StoragePool, networks []api.Network) (info *api.ClusterNodeAccepted, err error)
+	JoinCluster(targetAddress, targetPassword, targetCert, name string) (op *Operation, err error)
```

Comment:
I didn't look at this to deep. What you propose might work, except that there still needs to be an "Accept" and point which is different from "CreateClusterMember". That's mainly for internal use, because you need the joining node to hit an api on the target node to do some pre-work and return some info (such as if the joining node should be a database node). If we don't want to expose it we could probably remove AcceptNode and use RawQuery or something in the code.

Reply:
Actually, re-reading this now, the "Accept" endpoint would essentially be "CreateClusterMember" in your sketch.

At the moment BootstrapCluster() is pretty simple, it just performs a single POST /1.0/cluster/members request and it basically switches off the in-memory raft cluster and turns on the network version. We can easily rename it to UpdateCluster and change the endpoint to PUT /1.0/cluster.

On the other hand, JoinCluster is a bit more complex. Essentially it performs a request to the POST /1.0/cluster/members endpoint of the node that wants to join the cluster. Internally, the LXD API handler for that endpoint calls AcceptNode() against a client pointing to the target node of the cluster to be joined. AcceptNode() performs a request to POST /1.0/cluster/members, providing some details about the joining node (such as its networks and storage pools), and getting back some info from the target node (such as the private cert key, the existing database nodes, whether the joining node should promote itself to database node, etc).

If I understand you correctly, you're proposal would be to:

1. merge BootstrapCluster() and JoinCluster() into a single UpdateCluster() (changing the endpoint to PUT /1.0/cluster)

2. rename AcceptNode() to CreateClusterMember()

I believe 1. would be fine and not too difficult, as long as it's just a renaming/remix and we keep everything semantically unchanged, effectively (in particular the UpdateCluster API handler would still internally make that "accept" request described above).
 
One problem with 2. though is that this accept endpoint (however we call it) does not really create a cluster member per se, and probably should be considered somehow an internal endpoint.

So perhaps we could just drop AcceptNode()/CreateClusterMember() from the public API, leaving just GetCluster() and UpdateCluster(), moving the relevant logic somewhere else, and perhaps making use of RawOperation/RawQuery to hit the desired accept endpoint.

Hope it makes sense and it's clear.

Reply:
The new API layout and semantics are now in place.